File: java/13.roman-to-integer.java
/*
 * @lc app=leetcode id=13 lang=java
 *
 * [13] Roman to Integer
 */

// @lc code=start
class Solution {
    public int romanToInt(String s) {
        Map<Character, Integer> map = new HashMap<Character, Integer>(){{
            put('I',             1);
            put('V',             5);
            put('X',             10);
            put('L',             50);
            put('C',             100);
            put('D',             500);
            put('M',             1000);
        }};
        int res = 0, len=s.length();
        for(int i =0; i< len;i++){
            char cur = s.charAt(i);
            assert(map.containsKey(cur));
            int num = map.get(cur);
            if(i+1<len){
                char next = s.charAt(i+1);
                assert(map.containsKey(next));
                if(num < map.get(next))
                    num = -num;
            }
            res+= num;
            
            
        }
        return res;

        
    }
}
// @lc code=end

File: unsorted/js/13.roman-to-integer.js
/*
 * @lc app=leetcode id=13 lang=javascript
 *
 * [13] Roman to Integer
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */

var romanToInt = function(s) {
    const hash = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
    }
    let res = 0;
    for(let i = 0; i<s.length-1;i++){
        let ch =s[i];
        let next = s[i+1];
        let val1 = hash[ch];
        const val2 = hash[next];

        if(val1 < val2){
            val1 = -val1;
        }
        res += val1;

    }
    res += hash[s[s.length-1]];
    return res;
    
};
// @lc code=end

